pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/DefaultRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,333 @@
+      if (includeConsuming) {
+        rebalanceConsumingSegments(idealState, newPartitionAssignment);
+      }
+    }
+
+    // get target num replicas
+    int targetNumReplicas;
+    if (tableType.equals(CommonConstants.Helix.TableType.REALTIME)) {
+      String replicasString = tableConfig.getValidationConfig().getReplicasPerPartition();
+      try {
+        targetNumReplicas = Integer.parseInt(replicasString);
+      } catch (Exception e) {
+        throw new RuntimeException("Invalid value for replicasPerPartition:'" + replicasString + "'", e);
+      }
+    } else {
+      targetNumReplicas = Integer.parseInt(tableConfig.getValidationConfig().getReplication());
+    }
+
+    // always rebalance serving segments
+    rebalanceServingSegments(idealState, tableConfig, targetNumReplicas);
+
+    // update if not dryRun
+    boolean dryRun = rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.DRYRUN, DEFAULT_DRY_RUN);
+    if (!dryRun) {
+      LOGGER.info("Updating ideal state for table {}", tableNameWithType);
+      updateIdealState(tableNameWithType, targetNumReplicas, idealState.getRecord().getMapFields());
+    } else {
+      LOGGER.info("Dry run. Skip writing ideal state");
+    }
+    return idealState;
+  }
+
+  /**
+   * Rebalances consuming segments based on stream partition assignment
+   * @param idealState
+   * @param newPartitionAssignment
+   */
+  private void rebalanceConsumingSegments(IdealState idealState, PartitionAssignment newPartitionAssignment) {
+    LOGGER.info("Rebalancing consuming segments for table {}", idealState.getResourceName());
+
+    // for each partition, the segment with latest sequence number should be in CONSUMING
+    Map<Integer, LLCSegmentName> partitionIdToLatestSegment = new HashMap<>(newPartitionAssignment.getNumPartitions());
+    for (String segmentName : idealState.getPartitionSet()) {
+      LLCSegmentName llcSegmentName = new LLCSegmentName(segmentName);
+      int partitionId = llcSegmentName.getPartitionId();
+      LLCSegmentName latestSegmentForPartition = partitionIdToLatestSegment.get(partitionId);
+      if (latestSegmentForPartition == null
+          || llcSegmentName.getSequenceNumber() > latestSegmentForPartition.getSequenceNumber()) {
+        partitionIdToLatestSegment.put(partitionId, llcSegmentName);
+      }
+    }
+
+    // update ideal state of consuming segments, based on new stream partition assignment
+    Map<String, List<String>> partitionToInstanceMap = newPartitionAssignment.getPartitionToInstances();
+    for (LLCSegmentName llcSegmentName : partitionIdToLatestSegment.values()) {
+      int partitionId = llcSegmentName.getPartitionId();
+      Map<String, String> instanceStateMap = new HashMap<>();
+      for (String instance : partitionToInstanceMap.get(String.valueOf(partitionId))) {
+        instanceStateMap.put(instance, RealtimeSegmentOnlineOfflineStateModel.CONSUMING);
```

Comment:
you can add a comment here that it is possible that we turn an OFFLINE state into CONSUMING state in the process, but that should be OK since the only way a CONSUMING segment can become OFFLINE is when there are temporary or permanent issues consuming from the incoming event stream. If it turns back into CONSUMING, we will attempt to consume again (and may fail again if the fault is permanent)